File: src/sdk/agentraas/local.py
"""
Library mode: exactly-once execution with no AgentRaaS server at all.

    from agentraas.local import exactly_once

    @exactly_once()                       # SQLite file in the working dir
    def charge(customer, amount):
        return stripe.Charge.create(customer=customer, amount=amount).id

    charge("cus_1", 4200)   # runs for real
    charge("cus_1", 4200)   # retry: returns the cached id, no second charge

Same guarantee as the hosted proxy, same claim/complete/release model as
`crates/core/src/dedup.rs`: the first caller atomically claims a slot,
runs the function and stores its result; any retry or concurrent duplicate
gets that result back instead of running it again. If the function
raises, the slot is released so a retry can run it for real.

Stores: `SQLiteStore` (default, stdlib only, safe across threads and
processes on one machine) and `RedisStore` (needs `pip install redis`,
safe across machines). Results must be JSON-serializable.
"""

import asyncio
import functools
import hashlib
import inspect
import json
import sqlite3
import time

__all__ = ["exactly_once", "SQLiteStore", "RedisStore", "InFlightError"]

DEFAULT_TTL = 86400  # 24h, same as the server's default dedup window
_PENDING = '{"pending":true}'


class InFlightError(Exception):
    """A duplicate call arrived while the first one was still running and
    didn't finish within `wait` seconds. Retry later; it is not safe to
    assume the first call failed."""


class SQLiteStore:
    def __init__(self, path="agentraas-dedup.db"):
        self.path = path
        with self._conn() as c:
            c.execute("CREATE TABLE IF NOT EXISTS dedup (key TEXT PRIMARY KEY, value TEXT NOT NULL, expires_at REAL NOT NULL)")

    def _conn(self):
        # A connection per operation keeps this safe to share across threads.
        return sqlite3.connect(self.path, timeout=30, isolation_level=None)

    def claim(self, key, ttl):
        now = time.time()
        c = self._conn()
        try:
            c.execute("BEGIN IMMEDIATE")
            c.execute("DELETE FROM dedup WHERE key = ? AND expires_at < ?", (key, now))
            cur = c.execute("INSERT OR IGNORE INTO dedup VALUES (?, ?, ?)", (key, _PENDING, now + ttl))
            c.execute("COMMIT")
            return cur.rowcount == 1
        finally:
            c.close()

    def get(self, key):
        with self._conn() as c:
            row = c.execute("SELECT value FROM dedup WHERE key = ? AND expires_at >= ?", (key, time.time())).fetchone()
        return row[0] if row else None

    def complete(self, key, value, ttl):
        with self._conn() as c:
            c.execute("UPDATE dedup SET value = ?, expires_at = ? WHERE key = ?", (value, time.time() + ttl, key))

    def release(self, key):
        with self._conn() as c:
            c.execute("DELETE FROM dedup WHERE key = ?", (key,))


class RedisStore:
    def __init__(self, url="redis://localhost:6379/0", client=None):
        if client is None:
            import redis  # optional dependency
            client = redis.Redis.from_url(url)
        self.r = client

    def claim(self, key, ttl):
        return bool(self.r.set(key, _PENDING, nx=True, ex=ttl))

    def get(self, key):
        v = self.r.get(key)
        return v.decode() if isinstance(v, bytes) else v

    def complete(self, key, value, ttl):
        self.r.set(key, value, ex=ttl)

    def release(self, key):
        self.r.delete(key)


def _default_key(fn, args, kwargs):
    bound = inspect.signature(fn).bind_partial(*args, **kwargs)
    bound.apply_defaults()
    bound.arguments.pop("idempotency_key", None)
    return json.dumps(bound.arguments, sort_keys=True, default=str)


def exactly_once(store=None, key=None, ttl=DEFAULT_TTL, wait=30.0, poll=0.1):
    """Decorate a function (sync or async) so identical calls run once.

    key:  optional function taking the same arguments and returning the
          idempotency key (e.g. `key=lambda order, **_: order["id"]`).
          Default: all arguments, JSON-encoded.
    ttl:  how long a completed result is remembered, in seconds.
    wait: how long a duplicate waits for an in-flight first call.

    If the function has an `idempotency_key` parameter, it is filled with a
    key that stays the same across retries. Pass it to the provider (e.g.
    Stripe's Idempotency-Key header): that covers the one case this
    decorator can't, where the provider did the work but the response was
    lost, so the function raised and a retry runs it again.
    """
    store = store or SQLiteStore()

    def decorate(fn):
        name = f"{fn.__module__}.{fn.__qualname__}"
        wants_key = "idempotency_key" in inspect.signature(fn).parameters

        def slot(args, kwargs):
            raw = key(*args, **kwargs) if key else _default_key(fn, args, kwargs)
            return "dedup:local:" + hashlib.sha256(f"{name}\0{raw}".encode()).hexdigest()

        def inject(k, kwargs):
            if wants_key and kwargs.get("idempotency_key") is None:
                kwargs["idempotency_key"] = "ar-" + k[len("dedup:local:"):][:48]
            return kwargs

        def cached(k):
            v = store.get(k)
            if v is None or v == _PENDING:
                return v, None
            return v, json.loads(v)["result"]

        def finish(k, result):
            try:
                value = json.dumps({"result": result})
            except TypeError:
                store.release(k)
                raise TypeError(f"@exactly_once: {name} returned a non-JSON-serializable result; return plain data (ids, dicts)")
            store.complete(k, value, ttl)
            return result

        if inspect.iscoroutinefunction(fn):
            @functools.wraps(fn)
            async def async_wrapper(*args, **kwargs):
                k = slot(args, kwargs)
                deadline = time.monotonic() + wait
                while not store.claim(k, ttl):
                    v, result = cached(k)
                    if v not in (None, _PENDING):
                        return result
                    if time.monotonic() > deadline:
                        raise InFlightError(f"{name}: duplicate still in flight after {wait}s")
                    await asyncio.sleep(poll)
                try:
                    result = await fn(*args, **inject(k, kwargs))
                except BaseException:
                    store.release(k)
                    raise
                return finish(k, result)
            return async_wrapper

        @functools.wraps(fn)
        def wrapper(*args, **kwargs):
            k = slot(args, kwargs)
            deadline = time.monotonic() + wait
            while not store.claim(k, ttl):
                v, result = cached(k)
                if v not in (None, _PENDING):
                    return result
                if time.monotonic() > deadline:
                    raise InFlightError(f"{name}: duplicate still in flight after {wait}s")
                time.sleep(poll)
            try:
                result = fn(*args, **inject(k, kwargs))
            except BaseException:
                store.release(k)
                raise
            return finish(k, result)
        return wrapper

    return decorate

File: src/sdk-js/src/local.ts
/**
 * Library mode: exactly-once execution with no AgentRaaS server at all.
 *
 *   import { exactlyOnce, RedisStore } from "agentraas";
 *   import { createClient } from "redis";
 *
 *   const redis = await createClient().connect();
 *   const charge = exactlyOnce(
 *     async (customer: string, amount: number) => (await stripe.charges.create({ customer, amount })).id,
 *     { store: new RedisStore(redis), name: "charge" },
 *   );
 *   await charge("cus_1", 4200); // runs for real
 *   await charge("cus_1", 4200); // retry: cached id, no second charge
 *
 * Same claim/complete/release model as the hosted proxy: the first caller
 * claims a slot, runs the function and stores its result; retries and
 * concurrent duplicates get that result back. If the function throws, the
 * slot is released so a retry can run it for real. Results must be
 * JSON-serializable. `MemoryStore` is single-process only (tests, scripts);
 * use `RedisStore` for anything that restarts or runs more than one process.
 */
import { createHash } from "crypto";

const PENDING = '{"pending":true}';

/** Implement this to back exactlyOnce with any other database. */
export interface DedupStore {
  claim(key: string, ttlSeconds: number): Promise<boolean>;
  get(key: string): Promise<string | null>;
  complete(key: string, value: string, ttlSeconds: number): Promise<void>;
  release(key: string): Promise<void>;
}

export class InFlightError extends Error {}

export class MemoryStore implements DedupStore {
  private m = new Map<string, { v: string; exp: number }>();
  async claim(key: string, ttl: number) {
    const e = this.m.get(key);
    if (e && e.exp > Date.now()) return false;
    this.m.set(key, { v: PENDING, exp: Date.now() + ttl * 1000 });
    return true;
  }
  async get(key: string) {
    const e = this.m.get(key);
    return e && e.exp > Date.now() ? e.v : null;
  }
  async complete(key: string, v: string, ttl: number) {
    this.m.set(key, { v, exp: Date.now() + ttl * 1000 });
  }
  async release(key: string) {
    this.m.delete(key);
  }
}

/** Minimal slice of a node-redis v4+ client (`redis` package). */
interface RedisLike {
  set(key: string, value: string, opts: { NX?: boolean; EX?: number }): Promise<string | null>;
  get(key: string): Promise<string | null>;
  del(key: string): Promise<number>;
}

export class RedisStore implements DedupStore {
  constructor(private r: RedisLike) {}
  async claim(key: string, ttl: number) {
    return (await this.r.set(key, PENDING, { NX: true, EX: ttl })) === "OK";
  }
  get(key: string) {
    return this.r.get(key);
  }
  async complete(key: string, v: string, ttl: number) {
    await this.r.set(key, v, { EX: ttl });
  }
  async release(key: string) {
    await this.r.del(key);
  }
}

export interface ExactlyOnceOptions<A extends unknown[]> {
  store: DedupStore;
  /** Scopes keys; defaults to the function's own name. */
  name?: string;
  /** Idempotency key from the arguments. Default: all arguments, JSON-encoded. */
  key?: (...args: A) => string;
  /** How long a completed result is remembered. Default 24h. */
  ttlSeconds?: number;
  /** How long a duplicate waits for an in-flight first call. Default 30s. */
  waitMs?: number;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

/**
 * Wrap `fn` so identical calls run once. With `passKey: true`, `fn` gets a
 * retry-stable idempotency key as its first argument: pass it to the
 * provider (e.g. Stripe's `idempotencyKey` option). That covers the one
 * case the wrapper can't, where the provider did the work but the response
 * was lost, so `fn` threw and a retry runs it again.
 */
export function exactlyOnce<A extends unknown[], R>(
  fn: (...args: A) => Promise<R> | R,
  opts: ExactlyOnceOptions<A> & { passKey?: false },
): (...args: A) => Promise<R>;
export function exactlyOnce<A extends unknown[], R>(
  fn: (idempotencyKey: string, ...args: A) => Promise<R> | R,
  opts: ExactlyOnceOptions<A> & { passKey: true },
): (...args: A) => Promise<R>;

export function exactlyOnce<A extends unknown[], R>(
  fn: (...args: any[]) => Promise<R> | R,
  opts: ExactlyOnceOptions<A> & { passKey?: boolean },
): (...args: A) => Promise<R> {
  const name = opts.name ?? fn.name;
  if (!name) throw new Error("exactlyOnce: pass `name` for anonymous functions");
  const ttl = opts.ttlSeconds ?? 86400;
  const waitMs = opts.waitMs ?? 30000;
  const { store } = opts;

  return async (...args: A) => {
    const raw = opts.key ? opts.key(...args) : JSON.stringify(args);
    const k = "dedup:local:" + createHash("sha256").update(`${name}\0${raw}`).digest("hex");
    const deadline = Date.now() + waitMs;
    while (!(await store.claim(k, ttl))) {
      const v = await store.get(k);
      if (v !== null && v !== PENDING) return JSON.parse(v).result as R;
      if (Date.now() > deadline) throw new InFlightError(`${name}: duplicate still in flight after ${waitMs}ms`);
      await sleep(100);
    }
    let result: R;
    try {
      result = await (opts.passKey ? fn("ar-" + k.slice("dedup:local:".length, "dedup:local:".length + 48), ...args) : fn(...args));
    } catch (e) {
      await store.release(k);
      throw e;
    }
    const value = JSON.stringify({ result });
    if (value === undefined) {
      await store.release(k);
      throw new TypeError(`exactlyOnce: ${name} returned a non-JSON-serializable result`);
    }
    await store.complete(k, value, ttl);
    return result;
  };
}
